File: jvm/sargon-android/src/main/java/com/radixdlt/sargon/extensions/PublicKey.kt
package com.radixdlt.sargon.extensions

import com.radixdlt.sargon.BagOfBytes
import com.radixdlt.sargon.PublicKey
import com.radixdlt.sargon.ed25519PublicKeyToBytes
import com.radixdlt.sargon.ed25519PublicKeyToHex
import com.radixdlt.sargon.newEd25519PublicKeyFromBytes
import com.radixdlt.sargon.newEd25519PublicKeyFromHex
import com.radixdlt.sargon.newPublicKeyFromBytes
import com.radixdlt.sargon.newPublicKeyFromHex
import com.radixdlt.sargon.newSecp256k1PublicKeyFromBytes
import com.radixdlt.sargon.newSecp256k1PublicKeyFromHex
import com.radixdlt.sargon.publicKeyToBytes
import com.radixdlt.sargon.publicKeyToHex
import com.radixdlt.sargon.secp256k1PublicKeyToBytes
import com.radixdlt.sargon.secp256k1PublicKeyToBytesUncompressed
import com.radixdlt.sargon.secp256k1PublicKeyToHex

@Throws(SargonException::class)
fun PublicKey.Companion.init(hex: String): PublicKey = newPublicKeyFromHex(hex = hex)

@Throws(SargonException::class)
fun PublicKey.Companion.init(bytes: BagOfBytes): PublicKey =
    newPublicKeyFromBytes(bagOfBytes = bytes)

@Throws(SargonException::class)
fun PublicKey.Ed25519.Companion.init(hex: String): PublicKey.Ed25519 =
    PublicKey.Ed25519(newEd25519PublicKeyFromHex(hex = hex))

@Throws(SargonException::class)
fun PublicKey.Secp256k1.Companion.init(hex: String): PublicKey.Secp256k1 =
    PublicKey.Secp256k1(newSecp256k1PublicKeyFromHex(hex = hex))

@Throws(SargonException::class)
fun PublicKey.Ed25519.Companion.init(bytes: BagOfBytes): PublicKey.Ed25519 =
    PublicKey.Ed25519(newEd25519PublicKeyFromBytes(bytes = bytes))

@Throws(SargonException::class)
fun PublicKey.Secp256k1.Companion.init(bytes: BagOfBytes): PublicKey.Secp256k1 =
    PublicKey.Secp256k1(newSecp256k1PublicKeyFromBytes(bytes = bytes))

/**
 * Encodes the [PublicKey.Ed25519] to a hexadecimal string, lowercased, without any `0x` prefix,
 * e.g. `"b7a3c12dc0c8c748ab07525b701122b88bd78f600c76342d27f25e5f92444cde"`
 *
 * Encodes the compressed form (33 bytes) of a [PublicKey.Secp256k1] to a hexadecimal string, 
 * lowercased, without any `0x` prefix, e.g.
 * `"033083620d1596d3f8988ff3270e42970dd2a031e2b9b6488052a4170ff999f3e8"`
 */
val PublicKey.hex: String
    get() = publicKeyToHex(publicKey = this)

val PublicKey.bytes: BagOfBytes
    get() = publicKeyToBytes(publicKey = this)

val PublicKey.Ed25519.bytes: BagOfBytes
    get() = ed25519PublicKeyToBytes(publicKey = value)

/**
 * Returns the key on **compressed** form (33 bytes)
 */
val PublicKey.Secp256k1.bytes: BagOfBytes
    get() = secp256k1PublicKeyToBytes(publicKey = value)


/**
 * Returns the key on **uncompressed** form (65 bytes)
 *
 * Use `compressedData` for compressed format (33 bytes)
 */
val PublicKey.Secp256k1.uncompressedBytes: BagOfBytes
    get() = secp256k1PublicKeyToBytesUncompressed(publicKey = value)

File: crates/sargon-uniffi/src/core/types/keys/secp256k1/public_key.rs
use crate::{prelude::*, UniffiCustomTypeConverter};
use sargon::BagOfBytes as InternalBagOfBytes;
use sargon::Secp256k1PublicKey as InternalSecp256k1PublicKey;

/// A `secp256k1` public key used to verify cryptographic signatures (ECDSA signatures).
#[derive(Clone, PartialEq, Eq, Hash, uniffi::Record)]
pub struct Secp256k1PublicKey {
    secret_magic: BagOfBytes,
}

impl Secp256k1PublicKey {
    pub fn into_internal(&self) -> InternalSecp256k1PublicKey {
        self.clone().into()
    }
}

impl From<InternalSecp256k1PublicKey> for Secp256k1PublicKey {
    fn from(value: InternalSecp256k1PublicKey) -> Self {
        Self {
            secret_magic: value.to_bag_of_bytes().into(),
        }
    }
}

impl Into<InternalSecp256k1PublicKey> for Secp256k1PublicKey {
    fn into(self) -> InternalSecp256k1PublicKey {
        InternalSecp256k1PublicKey::try_from(self.secret_magic.to_vec()).unwrap()
    }
}

#[uniffi::export]
pub fn new_secp256k1_public_key_from_hex(
    hex: String,
) -> Result<Secp256k1PublicKey> {
    hex.parse::<InternalSecp256k1PublicKey>().into_result()
}

/// Creates a Secp256k1PublicKey from either compressed form (33 bytes) or
/// from uncompressed form (65 bytes).
#[uniffi::export]
pub fn new_secp256k1_public_key_from_bytes(
    bytes: BagOfBytes,
) -> Result<Secp256k1PublicKey> {
    InternalSecp256k1PublicKey::try_from(bytes.into_internal().to_vec())
        .into_result()
}

/// Encodes the compressed form (33 bytes) of a `Secp256k1PublicKey` to a hexadecimal string, lowercased, without any `0x` prefix, e.g.
/// `"033083620d1596d3f8988ff3270e42970dd2a031e2b9b6488052a4170ff999f3e8"`
#[uniffi::export]
pub fn secp256k1_public_key_to_hex(public_key: &Secp256k1PublicKey) -> String {
    public_key.into_internal().to_hex()
}

/// Returns the public key on **compressed** form (33 bytes)
#[uniffi::export]
pub fn secp256k1_public_key_to_bytes(
    public_key: &Secp256k1PublicKey,
) -> BagOfBytes {
    public_key.secret_magic.clone()
}

/// Returns the public key on **uncompressed** form (65 bytes)
#[uniffi::export]
pub fn secp256k1_public_key_to_bytes_uncompressed(
    public_key: &Secp256k1PublicKey,
) -> BagOfBytes {
    InternalBagOfBytes::from(public_key.into_internal().uncompressed()).into()
}

#[uniffi::export]
pub fn new_secp256k1_public_key_sample() -> Secp256k1PublicKey {
    InternalSecp256k1PublicKey::sample().into()
}

#[uniffi::export]
pub fn new_secp256k1_public_key_sample_other() -> Secp256k1PublicKey {
    InternalSecp256k1PublicKey::sample_other().into()
}
